lib/consumerGroup.js

```diff
@@ -0,0 +1,455 @@
+
+  this.recovery = new ConsumerGroupRecovery(this);
+
+  this.setupProtocols(this.options.protocol);
+
+  if (this.options.connectOnReady && !this.options.migrateHLC) {
+    this.client.once('ready', this.connect.bind(this));
+  }
+
+  if (this.options.migrateHLC) {
+    const ConsumerGroupMigrator = require('./consumerGroupMigrator');
+    this.migrator = new ConsumerGroupMigrator(this);
+  }
+
+  this.client.on('error', function (err) {
+    debug('Error from %s', self.client.clientId, err);
+    self.emit('error', err);
+  });
+
+  this.client.on('reconnect', function (lastError) {
+    self.fetch();
+  });
+
+  // 'done' will be emit when a message fetch request complete
+  this.on('done', function (topics) {
+    self.updateOffsets(topics);
+    if (!self.paused) {
+      setImmediate(function () {
+        self.fetch();
+      });
+    }
+  });
+
+  if (this.options.groupId) {
+    validateConfig('options.groupId', this.options.groupId);
+  }
+
+  this.isLeader = false;
+  this.coordinatorId = null;
+  this.generationId = null;
+  this.ready = false;
+  this.topicPayloads = [];
+}
+
+util.inherits(ConsumerGroup, highLevelConsumer);
+
+ConsumerGroup.prototype.setupProtocols = function (protocols) {
+  if (!Array.isArray(protocols)) {
+    protocols = [protocols];
+  }
+
+  this.protocols = protocols.map(function (protocol) {
+    if (typeof protocol === 'string') {
+      if (!(protocol in builtInProtocols)) {
+        throw new Error('Unknown built in assignment protocol ' + protocol);
+      }
+      protocol = _.assign({}, builtInProtocols[protocol]);
+    } else {
+      checkProtocol(protocol);
+    }
+
+    protocol.subscription = this.topics;
+    return protocol;
+  }, this);
+};
+
+function checkProtocol (protocol) {
+  assert(protocol, 'protocol is null');
+  assert(protocol.assign, 'assign function is not defined in the protocol');
+  assert(protocol.name, 'name must be given to protocol');
+  assert(protocol.version >= 0, 'version must be >= 0');
+}
+
+ConsumerGroup.prototype.setCoordinatorId = function (coordinatorId) {
+  this.client.coordinatorId = String(coordinatorId);
+};
+
+ConsumerGroup.prototype.assignPartitions = function (protocol, groupMembers, callback) {
+  debug('Assigning Partitions to members', groupMembers);
+  debug('Using group protocol', protocol);
+
+  protocol = _.find(this.protocols, {name: protocol});
+
+  var self = this;
+  var topics = _(groupMembers).pluck('subscription').flatten().uniq().value();
+
+  async.waterfall([
+    function (callback) {
+      debug('loadingMetadata for topics:', topics);
+      self.client.loadMetadataForTopics(topics, callback);
+    },
+
+    function (metadataResponse, callback) {
+      var metadata = mapTopicToPartitions(metadataResponse[1].metadata);
+      debug('mapTopicToPartitions', metadata);
+      callback(null, metadata);
+    },
+
+    function (metadata, callback) {
+      protocol.assign(metadata, groupMembers, callback);
+    }
+  ], callback);
```

Comment:
Is it possible that the waterfall method returns both and error and some of the data from protocol.assign? Maybe you should explicitly catch the error from the waterfall, check it and either ```return callback(null, data)``` or just ```return callback(error)``` to prevent any data leaks.